src/lib/tests/oauth.test.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * @jest-environment node
+ */
+import { ipcMain } from 'electron';
+import fs from 'fs';
+import { withMainWindow } from '../../main-window';
+import { loadUserData, saveUserData } from '../../storage/user-data';
+import { setUpAuthCallbackHandler } from '../oauth';
+
+jest.mock( 'fs' );
+jest.mock( '../../main-window' );
+jest.mock( '../../storage/user-data' );
+
+const mockUserData = {
+	sites: [],
+};
+( fs as MockedFs ).__setFileContents(
+	'/path/to/app/appData/App Name/appdata-v1.json',
+	JSON.stringify( mockUserData )
+);
+
+describe( 'setUpAuthCallbackHandler', () => {
+	it( 'should set up auth callback handler', async () => {
+		let authCallback: ( ...args: any[] ) => void = jest.fn();
+		( ipcMain.on as jest.Mock ).mockImplementationOnce( ( _event, callback ) => {
+			authCallback = callback;
+		} );
+		const mockSend = jest.fn();
+		( withMainWindow as jest.Mock ).mockImplementationOnce( ( callback ) => {
+			callback( {
+				webContents: {
+					send: mockSend,
+				},
+			} );
+		} );
+		const loadData = Promise.resolve( {} );
+		( loadUserData as jest.Mock ).mockReturnValueOnce( loadData );
+		const saveData = Promise.resolve( {} );
+		( saveUserData as jest.Mock ).mockReturnValueOnce( saveData );
+
+		setUpAuthCallbackHandler();
+		const mockToken = { email: 'mock-email' };
+		authCallback( null, { token: mockToken, error: null } );
+		await ( async () => {
+			await loadData;
+			await saveData;
+		} )();
```

Comment:
@fluiddot am I overlooking something obvious here? No rush on a reply.

Reply:
Sorry @dcalhoun for the delay on the response. I've taken a quick look and I think this is related to the following code involved in `setUpAuthCallbackHandler`:

- https://github.com/Automattic/studio/blob/12e236e16c1f5dde75278a92ef3f77cfbf5f500c/src/lib/oauth.ts#L114-L116
- https://github.com/Automattic/studio/blob/12e236e16c1f5dde75278a92ef3f77cfbf5f500c/src/lib/oauth.ts#L33-L43

Seems we need to explicitly wait for `loadUserData` and `saveUserData` promises before asserting that `mockSend` has been called with `auth-updated`. Alternative to this code, we could wait for promises to be resolved with:
```
await new Promise( process.nextTick );
```

```patch
diff --git forkSrcPrefix/src/lib/tests/oauth.test.ts forkDstPrefix/src/lib/tests/oauth.test.ts
index 865e0dae695785b356bc0fab0466bb6946f7b43c..44f9be6163fb3a4c3604ab3539e88c6715a6f867 100644
--- forkSrcPrefix/src/lib/tests/oauth.test.ts
+++ forkDstPrefix/src/lib/tests/oauth.test.ts
@@ -33,18 +33,13 @@ describe( 'setUpAuthCallbackHandler', () => {
 				},
 			} );
 		} );
-		const loadData = Promise.resolve( {} );
-		( loadUserData as jest.Mock ).mockReturnValueOnce( loadData );
-		const saveData = Promise.resolve( {} );
-		( saveUserData as jest.Mock ).mockReturnValueOnce( saveData );
+		( loadUserData as jest.Mock ).mockResolvedValueOnce( {} );
+		( saveUserData as jest.Mock ).mockResolvedValueOnce( {} );
 
 		setUpAuthCallbackHandler();
 		const mockToken = { email: 'mock-email' };
 		authCallback( null, { token: mockToken, error: null } );
-		await ( async () => {
-			await loadData;
-			await saveData;
-		} )();
+		await new Promise( process.nextTick );
 
 		expect( mockSend ).toHaveBeenCalledWith( 'auth-updated', { token: mockToken } );
 	} );

```

Reply:
> Seems we need to explicitly wait for `loadUserData` and `saveUserData` promises before asserting that `mockSend` has been called with `auth-updated`.

Correct, that is what I did with the IIFE. I was mostly confused as to why an IIFE was necessary, rather than merely awaiting the mocked `loadUserData` and `saveUserData` values inline. I feel I have successfully done the in the past. However, I realize now that the times I've leveraged that pattern have probably always been the in the context of React component testing, which likely operates differently with `act`. 

Regardless, the `await new Promise( process.nextTick );` approach is interesting, one that I had not considered before. Thanks for sharing that! 

I opened https://github.com/Automattic/studio/pull/107 simplifying the test case, hopefully improving its clarity.